File: archive/2019/solutions/SugarSampling.py
MOD = 1000000007

n = int(input())

numPaths = [0] * n * n
dp = [0] * n * n * n

numPaths[0] = 1
for i in range(n):
    a = [int(i)-1 for i in input().split()]
    for j in range(n):
        if i > 0:
            numPaths[n*i+j] = (numPaths[n*i+j] + numPaths[n*(i-1)+j]) % MOD
        if j > 0:
            numPaths[n*i+j] = (numPaths[n*i+j] + numPaths[n*i+j-1]) % MOD
        for k in range(n):
            if a[j] == k:
                dp[n*n*i+n*j+k] = numPaths[n*i+j]
            else:
                if i > 0:
                    dp[n*n*i+n*j+k] = (dp[n*n*i+n*j+k] + dp[n*n*(i-1)+n*j+k]) % MOD
                if j > 0:
                    dp[n*n*i+n*j+k] = (dp[n*n*i+n*j+k] + dp[n*n*i+n*(j-1)+k]) % MOD

for i in range(n):
    print(i+1, dp[n*n*(n-1)+n*(n-1)+i])

File: archive/2019/solutions/SugarSampling.cpp
#include <bits/stdc++.h>
using namespace std;

#define MAXN 201
#define MOD 1000000007

int main() {
    int n, a[MAXN][MAXN], numPaths[MAXN][MAXN] = {1}, dp[MAXN][MAXN][MAXN];
    cin >> n;
    for (int i=0; i<n; i++)
        for (int j=0; j<n; j++) {
            cin >> a[i][j];
            a[i][j]--;
            if (i > 0) numPaths[i][j] = (numPaths[i][j] + numPaths[i-1][j]) % MOD;
            if (j > 0) numPaths[i][j] = (numPaths[i][j] + numPaths[i][j-1]) % MOD;
            for (int k=0; k<n; k++) {
                if (a[i][j] == k)
                    dp[i][j][k] = numPaths[i][j];
                else {
                    if (i > 0) dp[i][j][k] = (dp[i][j][k] + dp[i-1][j][k]) % MOD;
                    if (j > 0) dp[i][j][k] = (dp[i][j][k] + dp[i][j-1][k]) % MOD;
                }
            }
        }

    for (int i=0; i<n; i++)
        cout << i+1 << " " << dp[n-1][n-1][i] << endl;

    return 0;
}